File: App/app/src/main/java/com/cinema/ticket_booking/di/AppModule.java
package com.cinema.ticket_booking.di;

import android.content.Context;
import androidx.room.Room;
import com.cinema.ticket_booking.data.local.AppDatabase;
import com.cinema.ticket_booking.data.local.TokenManager;
import com.cinema.ticket_booking.data.local.dao.BookingDao;
import com.cinema.ticket_booking.data.local.dao.MovieDao;
import com.cinema.ticket_booking.data.repository.*;
import com.cinema.ticket_booking.network.ApiService;
import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;
import javax.inject.Singleton;

@Module
@InstallIn(SingletonComponent.class)
public class AppModule {

    @Provides @Singleton
    public TokenManager provideTokenManager(@ApplicationContext Context ctx) {
        return new TokenManager(ctx);
    }

    @Provides @Singleton
    public AppDatabase provideDatabase(@ApplicationContext Context ctx) {
        return Room.databaseBuilder(ctx, AppDatabase.class, "cinema_db")
                .fallbackToDestructiveMigration().build();
    }

    @Provides @Singleton
    public MovieDao provideMovieDao(AppDatabase db) { return db.movieDao(); }

    @Provides @Singleton
    public BookingDao provideBookingDao(AppDatabase db) { return db.bookingDao(); }

    @Provides @Singleton
    public AuthRepository provideAuthRepository(ApiService api, TokenManager tm) {
        return new AuthRepository(api, tm);
    }

    @Provides @Singleton
    public MovieRepository provideMovieRepository(ApiService api) {
        return new MovieRepository(api);
    }

    @Provides @Singleton
    public BookingRepository provideBookingRepository(ApiService api) {
        return new BookingRepository(api);
    }

    @Provides @Singleton
    public ShowtimeRepository provideShowtimeRepository(ApiService api) {
        return new ShowtimeRepository(api);
    }

    @Provides @Singleton
    public CinemaRepository provideCinemaRepository(ApiService api) {
        return new CinemaRepository(api);
    }

    @Provides @Singleton
    public VoucherRepository provideVoucherRepository(ApiService api) {
        return new VoucherRepository(api);
    }

    @Provides @Singleton
    public ComboRepository provideComboRepository(ApiService api) {
        return new ComboRepository(api);
    }

    @Provides @Singleton
    public UserRepository provideUserRepository(ApiService api) {
        return new UserRepository(api);
    }

    @Provides @Singleton
    public NotificationRepository provideNotificationRepository(ApiService api) {
        return new NotificationRepository(api);
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/di/AppModule.kt
package com.cinema.ticket_booking.di

import android.content.Context
import androidx.room.Room
import com.cinema.ticket_booking.data.local.AppDatabase
import com.cinema.ticket_booking.data.local.TokenManager
import com.cinema.ticket_booking.data.local.dao.BookingDao
import com.cinema.ticket_booking.data.local.dao.MovieDao
import com.cinema.ticket_booking.data.repository.*
import com.cinema.ticket_booking.network.ApiService
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object AppModule {

    @Provides
    @Singleton
    fun provideTokenManager(@ApplicationContext ctx: Context): TokenManager {
        return TokenManager(ctx)
    }

    @Provides
    @Singleton
    fun provideDatabase(@ApplicationContext ctx: Context): AppDatabase {
        return Room.databaseBuilder(ctx, AppDatabase::class.java, "cinema_db")
            .fallbackToDestructiveMigration().build()
    }

    @Provides
    @Singleton
    fun provideMovieDao(db: AppDatabase): MovieDao {
        return db.movieDao()
    }

    @Provides
    @Singleton
    fun provideBookingDao(db: AppDatabase): BookingDao {
        return db.bookingDao()
    }

    @Provides
    @Singleton
    fun provideAuthRepository(api: ApiService, tm: TokenManager): AuthRepository {
        return AuthRepository(api, tm)
    }

    @Provides
    @Singleton
    fun provideMovieRepository(api: ApiService): MovieRepository {
        return MovieRepository(api)
    }

    @Provides
    @Singleton
    fun provideBookingRepository(api: ApiService, bookingDao: BookingDao): BookingRepository {
        return BookingRepository(api, bookingDao)
    }

    @Provides
    @Singleton
    fun provideShowtimeRepository(api: ApiService): ShowtimeRepository {
        return ShowtimeRepository(api)
    }

    @Provides
    @Singleton
    fun provideCinemaRepository(api: ApiService): CinemaRepository {
        return CinemaRepository(api)
    }

    @Provides
    @Singleton
    fun provideVoucherRepository(api: ApiService): VoucherRepository {
        return VoucherRepository(api)
    }

    @Provides
    @Singleton
    fun provideComboRepository(api: ApiService): ComboRepository {
        return ComboRepository(api)
    }

    @Provides
    @Singleton
    fun provideUserRepository(api: ApiService): UserRepository {
        return UserRepository(api)
    }

    @Provides
    @Singleton
    fun provideNotificationRepository(api: ApiService): NotificationRepository {
        return NotificationRepository(api)
    }

    @Provides
    @Singleton
    fun provideStaffRepository(api: ApiService): StaffRepository {
        return StaffRepository(api)
    }
}
